master/buildbot/reporters/words.py

```diff
@@ -678,369 +773,473 @@ def command_FORCE(self, args):
+            self.send("Stopping buildbot.")
+            self.bot.reactor.stop()
     command_SHUTDOWN.usage = {
         None: "shutdown check|start|stop|now - shutdown the buildbot master",
         "check": "shutdown check - check if the buildbot master is running or shutting down",
         "start": "shutdown start - start a clean shutdown",
         "stop": "shutdown cancel - stop the clean shutdown",
         "now": "shutdown now - shutdown immediately without waiting for the builders to finish"}
 
-    # communication with the user
 
-    def send(self, message):
-        if self.muted:
-            return
+class StatusBot(service.AsyncMultiService):
 
-        if self.channel:
-            self.bot.groupChat(self.channel, message)
+    """ Abstract status bot """
+
+    contactClass = Contact
+    channelClass = Channel
+
+    commandPrefix = ''
+    commandSuffix = None
+
+    offline_string = "offline"
+    idle_string = "idle"
+    running_string = "running:"
+
+    def __init__(self, authz=None, tags=None, notify_events=None,
+                 useRevisions=False, showBlameList=False):
+        super().__init__()
+        self.tags = tags
+        if notify_events is None:
+            notify_events = {}
+        self.notify_events = notify_events
+        self.useRevisions = useRevisions
+        self.showBlameList = showBlameList
+        self.authz = self.expand_authz(authz)
+        self.contacts = {}
+        self.channels = {}
+        self.reactor = reactor
+
+    @staticmethod
+    def expand_authz(authz):
+        if authz is None:
+            authz = {}
+        expanded_authz = {}
+        for cmds, val in authz.items():
+            if not isinstance(cmds, (tuple, list)):
+                cmds = (cmds,)
+            for cmd in cmds:
+                expanded_authz[cmd.upper()] = val
+        return expanded_authz
+
+    def getContact(self, user, channel=None):
+        """ get a Contact instance for ``user`` on ``channel`` """
+        try:
+            return self.contacts[(channel, user)]
+        except KeyError:
+            new_contact = self.contactClass(self, user=user, channel=channel)
+            self.contacts[(channel, user)] = new_contact
+            return new_contact
+
+    def getChannel(self, channel):
+        try:
+            return self.channels[channel]
+        except KeyError:
+            new_channel = self.channelClass(self, channel)
+            self.channels[channel] = new_channel
+            new_channel.setServiceParent(self)
+            return new_channel
+
+    def _get_object_id(self):
+        return self.master.db.state.getObjectId(
+            self.nickname, '{0.__module__}.{0.__name__}'.format(self.__class__))
+
+    @defer.inlineCallbacks
+    def _save_channels_state(self, attr, json_type=None):
+        if json_type is None:
+            json_type = lambda x: x
+        data = [(channel.id, json_type(getattr(channel, attr)))
+                for channel in self.channels.values()]
```

Comment:
If I understand correctly, this will grow without limit even if notification is disabled, right?